proxy/http/HttpConnectionCount.cc

```diff
@@ -21,51 +21,318 @@
+  Ticker last_tick{_last_alert};                  // Load the most recent alert time in ticks.
+  TimePoint last{TimePoint::duration{last_tick}}; // Most recent alert time in a time_point.
+  TimePoint now = Clock::now();                   // Current time_point.
+  if (last + _global_config->alert_delay <= now) {
+    // it's been long enough, swap out our time for the last time. The winner of this swap
+    // does the actual alert, leaving its current time as the last alert time.
+    zret = _last_alert.compare_exchange_strong(last_tick, now.time_since_epoch().count());
+    if (zret && lat) {
+      *lat = Clock::to_time_t(last);
+    }
+  }
+  return zret;
+}
+
+std::time_t
+OutboundConnTrack::Group::get_last_alert_epoch_time() const
+{
+  return Clock::to_time_t(TimePoint{TimePoint::duration{Ticker{_last_alert}}});
+}
+
+void
+OutboundConnTrack::get(std::vector<Group const *> &groups)
+{
+  std::lock_guard<std::mutex> lock(_imp._mutex); // TABLE LOCK
+  groups.resize(0);
+  groups.reserve(_imp._table.count());
+  for (Group const &g : _imp._table) {
+    groups.push_back(&g);
+  }
+}
 
 std::string
-ConnectionCount::dumpToJSON()
+OutboundConnTrack::to_json_string()
 {
-  Vec<ConnAddr> keys;
-  ink_mutex_acquire(&_mutex);
-  _hostCount.get_keys(keys);
-  std::ostringstream oss;
-  oss << '{';
-  appendJSONPair(oss, "connectionCountSize", keys.n);
-  oss << ", \"connectionCountList\": [";
-  for (size_t i = 0; i < keys.n; i++) {
-    oss << '{';
+  std::string text;
+  size_t extent = 0;
+  static const ts::BWFormat header_fmt{R"({{"count": {}, "list": [
+)"};
+  static const ts::BWFormat item_fmt{
+    R"(  {{"type": "{}", "ip": "{}", "fqdn": "{}", "current": {}, "max": {}, "blocked": {}, "queued": {}, "alert": {}}},
+)"};
+  static const std::string_view trailer{"]}"};
+
+  static const auto printer = [](ts::BufferWriter &w, Group const *g) -> ts::BufferWriter & {
+    w.print(item_fmt, g->_match_type, g->_addr, g->_fqdn, g->_count.load(), g->_count_max.load(), g->_blocked.load(),
+            g->_rescheduled.load(), g->get_last_alert_epoch_time());
+    return w;
+  };
 
-    appendJSONPair(oss, "ip", keys[i].getIpStr());
-    oss << ',';
+  ts::FixedBufferWriter null_bw{nullptr}; // Empty buffer for sizing work.
+  std::vector<Group const *> groups;
 
-    appendJSONPair(oss, "port", keys[i]._addr.host_order_port());
-    oss << ',';
+  self_type::get(groups);
 
-    appendJSONPair(oss, "hostname_hash", keys[i].getHostnameHashStr());
-    oss << ',';
+  extent += null_bw.print(header_fmt, groups.size()).extent();
+  for (auto g : groups) {
+    extent += printer(null_bw, g).extent();
+  }
+  extent += trailer.size();
+
+  text.resize(extent);
+  ts::FixedBufferWriter w(const_cast<char *>(text.data()), text.size());
+  w.print(header_fmt, groups.size());
+  for (auto g : groups) {
+    printer(w, g);
+  }
+  if (groups.size() > 0 && w.remaining()) {
+    w.auxBuffer()[-2] = ' '; // convert trailing comma to space.
```

Comment:
Really?  Really?  This is why we can't have nice abstractions.

Reply:
Yeah, it's a bit ugly. I'm trying to decide how to do this better. But the buffer is known to be long enough so this works on a `LocalBufferWriter`. I'm open to suggestions.

Reply:
Aren't you missing a "fill()" call here anyway?  Can't you just make size of the fill() one less to not include the dangling comma?

Reply:
No. There is no missing call to `fill`. I wonder if it might work to `clip(2)` and then paste the terminator in. Have to try that. The `std::string` has been resized to be the exact expected output length at that point.

Reply:
Hmmm not seeing that.